From 07be32144041d466d2d9016c6a26cf89ecdc36be Mon Sep 17 00:00:00 2001
From: Joshua Wilkinson <joshua.wilkinson@morsemicro.com>
Date: Tue, 15 Apr 2025 00:59:38 +0000
Subject: [PATCH 4/7] Merged in SW-14454-send-retry-exch-req-to-peer-mac (pull
 request #283)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

SW-14454 - Send PKEX Exchange Request retries to the peer_mac, rather than the broadcast address

Currently, hostapd has provision for sending 5 additional retries of the PKEX Exchange Request message to an Enrollee during DPP-PB. The implementation sends the first Exchange Request (in hostapd_dpp_pb_pkex_init) as a directed frame to the peer identified for DPP, but the retries are transmitted as broadcast.

According to the Wi-Fi Easy Connect Specification v3.0, Section 5.6.1 PKEX Preliminaries (pp. 71):
```
The Initiator shall iterate through the list of channels on the supported bands. If the Initiator of PKEX has learned the
MAC address of the Responder in a manner outside the scope of this specification, it shall send its PKEX Exchange
request to that MAC address, otherwise, it shall send its PKEX Exchange Request to the broadcast MAC address.
```
The peers MAC address has been discovered through the Push-Button announcements, so I would argue these frames should be directed. Note that for other methods, namely DPP over TCP, we don’t have that information, so we fall back to the broadcast address instead.

Approved-by: Andrew Pope
Approved-by: Nev Young
Approved-by: Simon Wadsworth
Approved-by: Adam Baumann
Approved-by: Sam Chapman
Approved-by: James Herbert
Approved-by: Feddrick Aquino
---
 src/ap/dpp_hostapd.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/src/ap/dpp_hostapd.c b/src/ap/dpp_hostapd.c
index 075091c95..1556b1ffd 100644
--- a/src/ap/dpp_hostapd.c
+++ b/src/ap/dpp_hostapd.c
@@ -466,6 +466,7 @@ static void hostapd_dpp_pkex_retry_timeout(void *eloop_ctx, void *timeout_ctx)
 {
 	struct hostapd_data *hapd = eloop_ctx;
 	struct dpp_pkex *pkex = hapd->dpp_pkex;
+	const u8 *dst;
 
 	if (!pkex || !pkex->exchange_req)
 		return;
@@ -490,16 +491,16 @@ static void hostapd_dpp_pkex_retry_timeout(void *eloop_ctx, void *timeout_ctx)
 	}
 
 	pkex->exch_req_tries++;
+	dst = pkex->peer_mac ? pkex->peer_mac : broadcast;
 	wpa_printf(MSG_DEBUG, "DPP: Retransmit PKEX Exchange Request (try %u)",
 		   pkex->exch_req_tries);
 	wpa_msg(hapd->msg_ctx, MSG_INFO, DPP_EVENT_TX "dst=" MACSTR
 		" freq=%u type=%d",
-		MAC2STR(broadcast), pkex->freq,
+		MAC2STR(dst), pkex->freq,
 		pkex->v2 ? DPP_PA_PKEX_EXCHANGE_REQ :
 		DPP_PA_PKEX_V1_EXCHANGE_REQ);
 	hostapd_drv_send_action(hapd, pkex->freq, pkex->exch_req_wait_time,
-				broadcast,
-				wpabuf_head(pkex->exchange_req),
+				dst, wpabuf_head(pkex->exchange_req),
 				wpabuf_len(pkex->exchange_req));
 }
 
@@ -2494,6 +2495,7 @@ static void hostapd_dpp_pb_pkex_init(struct hostapd_data *hapd,
 		return;
 	}
 	pkex->freq = freq;
+	os_memcpy(pkex->peer_mac, src, ETH_ALEN);
 
 	hapd->dpp_pkex = pkex;
 	msg = hapd->dpp_pkex->exchange_req;
-- 
2.34.1

